thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java

```diff
@@ -15,53 +15,226 @@
+ * is about to finish to free its presenter. For example when the Presenter is in the backstack.
+ * This savior uses {@link android.app.Application.ActivityLifecycleCallbacks} to detect those
+ * cases
+ * and destroys and cleans the presenters to prevents leaks.
+ * </p>
  */
-public enum PresenterSavior implements TiPresenterSavior {
+public class PresenterSavior implements TiPresenterSavior, ActivityInstanceObserver.Listener {
 
-    INSTANCE;
+    private static PresenterSavior INSTANCE;
 
     private static final String TAG = PresenterSavior.class.getSimpleName();
 
-    private final HashMap<String, TiPresenter> mPresenters = new HashMap<>();
+    @VisibleForTesting
+    static final String TI_ACTIVITY_PRESENTER_SCOPE_KEY = "ThirtyInch_presenter_scope_id";
+
+    /**
+     * enable debug logging for testing
+     */
+    private static final boolean DEBUG = false;
+
+    @VisibleForTesting
+    ActivityInstanceObserver mActivityInstanceObserver;
+
+    /**
+     * Holds a scope for every Activity with one or more presenters. There is no direct mapping for
+     * {@link Activity} to {@link ActivityScopedPresenters} because Activity instances can be
+     * destroyed. The {@link ActivityInstanceObserver} takes care to manage unique Ids for each
+     * Activity which are used as keys here.
+     */
+    @VisibleForTesting
+    final HashMap<String, ActivityScopedPresenters> mScopes = new HashMap<>();
+
+    /**
+     * Access to the {@link PresenterSavior} singleton to save presenters across orientation changes
+     *
+     * @return singleton for
+     */
+    public static synchronized PresenterSavior getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new PresenterSavior();
+        }
+        return INSTANCE;
+    }
+
+    @VisibleForTesting
+    PresenterSavior() {
+
+    }
+
+    @Override
+    public void free(final String presenterId, @NonNull final Activity activity) {
+        final ActivityScopedPresenters scope = getScope(activity);
+        if (scope != null) {
+            scope.remove(presenterId);
+            if (scope.isEmpty()) {
+                String scopeId = mActivityInstanceObserver.getActivityId(activity);
+                if (scopeId != null) {
+                    mScopes.remove(scopeId);
+                }
+            }
+        }
+        if (mScopes.isEmpty()) {
+            if (mActivityInstanceObserver != null) {
+                activity.getApplication()
+                        .unregisterActivityLifecycleCallbacks(mActivityInstanceObserver);
+                mActivityInstanceObserver = null;
+            }
+        }
+
+        printRemainingPresenter();
+    }
 
     @Override
-    public void free(final String presenterId) {
-        mPresenters.remove(presenterId);
+    public void onActivityFinished(final Activity activity, final String activityId) {
+        final ActivityScopedPresenters scope = mScopes.remove(activityId);
```

Comment:
the scope is only required in on branch, true. But removing the scope is the important part here. required for both branches.